nmrfx-processor/src/main/java/org/nmrfx/processor/datasets/vendor/JCAMPData.java

```diff
@@ -17,987 +17,569 @@
+    public void close() {
+        // Nothing to close
     }
 
     @Override
-    public Map getFidFlags() {
-        Map<String, Boolean> flags = new HashMap(5);
-        return flags;
+    public String getFilePath() {
+        return path;
     }
 
     @Override
-    public String getFilePath() {
-        return fpath;
+    public String getPar(String parname) {
+        return block.get(parname).getString();
     }
 
     @Override
-    public DatasetType getPreferredDatasetType() {
-        return preferredDatasetType;
+    public Double getParDouble(String parname) {
+        return block.get(parname).getDouble();
     }
 
     @Override
-    public void setPreferredDatasetType(DatasetType datasetType) {
-        this.preferredDatasetType = datasetType;
+    public Integer getParInt(String parname) {
+        return block.get(parname).getInt();
     }
 
     @Override
     public List<VendorPar> getPars() {
+        Set<String> defined = new HashSet<>();
         List<VendorPar> vendorPars = new ArrayList<>();
-        for (Map.Entry<String, String> par : parMap.entrySet()) {
-            vendorPars.add(new VendorPar(par.getKey(), par.getValue()));
+
+        // get block-level records first
+        for (String key : block.allRecordKeys()) {
+            if (defined.add(key)) {
+                vendorPars.add(new VendorPar(key, block.get(key).getString()));
+            }
+        }
+
+        // then add document-level records if they were not defined by the block
+        for (String key : document.allRecordKeys()) {
+            if (defined.add(key)) {
+                vendorPars.add(new VendorPar(key, document.get(key).getString()));
+            }
         }
+
         return vendorPars;
     }
 
     @Override
-    public String getPar(String parname) {
-        if (parMap == null) {
-            return null;
-        } else {
-            return parMap.get(parname);
-        }
+    public int getNVectors() {
+        //XXX To confirm, looks like it is number of 1D rows in total
+        // so 1 if 1D data, and number of 2D if 2D data.
+        // no more than 2D data are supported in JCamp
+
+        //XXX should we try to get it from page content instead?
+        return block.optional($TD, 1)
+                .map(r -> r.getInt() / 2)
+                .orElse(1);
     }
 
     @Override
-    public Double getParDouble(String parname) {
-        if ((parMap == null) || (parMap.get(parname) == null)) {
-            return null;
-//            throw new NullPointerException();
-        } else {
-            return Double.parseDouble(parMap.get(parname));
-        }
+    public int getNPoints() {
+        // XXX original JCAMPData always returned 0 (np / 2, with np never initialized). Isn't it a bug?
+
+        // XXX always /2, or only if complex data?
+        // XXX should we try to get it from page content instead?
+        return block.optional($TD, 0)
+                .map(r -> r.getInt() / 2)
+                .orElse(0);
```

Comment:
This is dubious as well. I'm pretty sure the original had a bug with `np` never initialized.

Reply:
Yes, that was a bug, though it looks like getNPoints wasn't actually used here.  The size was determined in readVEctor from the number of yValues parsed